{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import sys, os\n",
    "import time, pickle\n",
    "import re\n",
    "from ete3 import NCBITaxa\n",
    "sys.path.append(\"/Users/chilpert/Work/pyproteinsExt/src\")\n",
    "sys.path.append(\"/Users/chilpert/Work/pyproteins/src\")\n",
    "import pyproteinsExt.topology as topology\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save(data, tag=None):\n",
    "    print(data)\n",
    "    saveDir=\"/Volumes/arwen/mobi/group/NOX_CH/pickle_saved\"\n",
    "    timestr = time.strftime(\"%Y%m%d-%H%M%S\")\n",
    "    fTag = \"NOX_annotation_\" + tag + \"_\" if tag else \"NOX_annotation_\"\n",
    "    fSerialDump = fTag + timestr + \".pickle\"\n",
    "    with open(saveDir + '/' + fSerialDump, 'wb') as f:\n",
    "        pickle.dump(data, f)\n",
    "    print('data structure saved to', saveDir + '/' + fSerialDump)\n",
    "\n",
    "def load(fileName):\n",
    "    saveDir=\"/Volumes/arwen/mobi/group/NOX_CH/pickle_saved\"\n",
    "    d = pickle.load( open(saveDir + \"/\" + fileName, \"rb\" ) )\n",
    "    print(\"restore a annotated container of \", len(d), \"elements\")\n",
    "    return d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Raw Pfam domains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, NOX signature search is done by using Pfam domains corresponding to FAD_binding, NAD_binding and Ferric_reduct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preliminary work \n",
    "Starting from complete trEMBL dataset `arwen:/mobi/group/NOX_CH/data/uniprot_trembl.fasta.gz` which is a symbolic link for `arwen:/mobi/group/databases/flat/uniprot_trembl_2019_02.fasta.gz`\n",
    " *  Split the dataset in small volumes\n",
    "     * script: <span style=\"color:green\">**split.py**</span>\n",
    "     * Usage:\n",
    "     Create and go to the `/mobi/group/NOX_GH/volumes` \n",
    "```console\n",
    "    ROOT_DIR=/mobi/group/NOX_CH\n",
    "    SCRIPT_DIR=/mobi/group/NOX_CH/nox-analysis/scripts\n",
    "    $SCRIPT_DIR/split.py $ROOT_DIR/data/uniprot_trembl.fasta.gz\n",
    "```\n",
    "\n",
    "* Run the HMMR and TMHMM annotations\n",
    "    * script: <span style=\"color:green\">**runHMMR_slurm.sh**</span>\n",
    "    * Usage:  \n",
    "  \n",
    "```console\n",
    "    mkdir $ROOT_DIR/seedSet\n",
    "    mkdir $ROOT_DIR/seedSet/work\n",
    "    $SCRIPT_DIR/runHMMR_slurm.sh $ROOT_DIR/volumes $ROOT_DIR/seedSet/work $ROOT_DIR/data/profiles\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataDir=glob.glob('/Volumes/arwen/mobi/group/NOX_CH/seedSet/work/uniprot_trembl_v*')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parse hmmr, tmhmm and fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parsing_results(dataDir):\n",
    "    dataDir_length=len(dataDir)\n",
    "    c=1\n",
    "    print(c,\"/\",dataDir_length)\n",
    "    dataContainer=topology.parse(hmmrOut=dataDir[0]+\"/hmmsearch.out\",tmhmmOut=dataDir[0]+\"/tmhmm.out\",fastaOut=dataDir[0]+\"/hmmsearch.fasta\")\n",
    "    for dir in dataDir[1:]:\n",
    "        c+=1\n",
    "        if c%10==0:\n",
    "            print(c,\"/\",dataDir_length)\n",
    "        dataContainer=dataContainer.addParsing(topology.parse(hmmrOut=dir+\"/hmmsearch.out\",tmhmmOut=dir+\"/tmhmm.out\",fastaOut=dir+\"/hmmsearch.fasta\"))\n",
    "    return dataContainer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 / 293\n",
      "10 / 293\n",
      "20 / 293\n",
      "30 / 293\n",
      "40 / 293\n",
      "50 / 293\n",
      "60 / 293\n",
      "70 / 293\n",
      "80 / 293\n",
      "90 / 293\n",
      "100 / 293\n",
      "110 / 293\n",
      "120 / 293\n",
      "130 / 293\n",
      "140 / 293\n",
      "150 / 293\n",
      "160 / 293\n",
      "170 / 293\n",
      "180 / 293\n",
      "190 / 293\n",
      "200 / 293\n",
      "210 / 293\n",
      "220 / 293\n",
      "230 / 293\n",
      "240 / 293\n",
      "250 / 293\n",
      "260 / 293\n",
      "Warning:: >> tr|A0A0F0DDW2|A0A0F0DDW2_9PSED  Cytochrome B561 OS=Pseudomonas sp. 21 OX=1619948 GN=UB43_27990 PE=4 SV=1\n",
      "   [No individual domains that satisfy reporting thresholds (although complete target did)]\n",
      "\n",
      "\n",
      "270 / 293\n",
      "280 / 293\n",
      "290 / 293\n"
     ]
    }
   ],
   "source": [
    "dataContainer=parsing_results(dataDir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins : 178540\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of proteins :\",len(dataContainer))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<pyproteinsExt.topology.TopologyContainer object at 0x117480710>\n",
      "data structure saved to /Volumes/arwen/mobi/group/NOX_CH/pickle_saved/NOX_annotation_All_20190613-162757.pickle\n"
     ]
    }
   ],
   "source": [
    "save(dataContainer,\"All\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter proteins that match the three domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_three_domains(entry): \n",
    "    domains=set([hmm_obj.domain for hmm_obj in entry.hmmr])\n",
    "    if len(domains)==3:\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredThreeDomains=dataContainer.filter(filter_three_domains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins with 3 domains :  18020\n"
     ]
    }
   ],
   "source": [
    "print('Number of proteins with 3 domains : ',len(filteredThreeDomains))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter helix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Helix number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_nb_helix(entry,**kwargs):\n",
    "    min_helix=kwargs[\"min_helix\"]\n",
    "    max_helix=kwargs[\"max_helix\"]\n",
    "    if entry.tmhmm.nb_helix>=min_helix and entry.tmhmm.nb_helix<=max_helix:\n",
    "        return True \n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredNbHelix=filteredThreeDomains.filter(filter_nb_helix,min_helix=2,max_helix=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins with 3 domains and between 2 and 7 helix : 15559\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of proteins with 3 domains and between 2 and 7 helix :\",len(filteredNbHelix.entries))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Bi-histidine "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_bi_histidine(entry):\n",
    "    seq=entry.fasta.seq\n",
    "    topo_seq=entry.tmhmm.topology_seq\n",
    "    H_status = []\n",
    "    if len(topo_seq)!=len(seq):\n",
    "        raise Exception(\"Topology seq and amino acids seq with different size. Check !\")\n",
    "    for i in range(len(seq)):\n",
    "        if topo_seq[i]==\"i\" or topo_seq[i]==\"o\":\n",
    "            continue\n",
    "        if not seq[i]==\"H\":\n",
    "            continue\n",
    "        H_status.append( [i, topo_seq[i], False] )     \n",
    "    for i in range (len(H_status) - 1):\n",
    "        for j in range (i + 1, len(H_status)):\n",
    "            if H_status[i][1] != H_status[j][1]:\n",
    "                continue\n",
    "            d = H_status[i][0] - H_status[j][0]\n",
    "            if d >= 12 or d <= 14:\n",
    "                H_status[i][2] = True\n",
    "                \n",
    "    # Only keep marked histidine\n",
    "    H_status = [ x for x in H_status if x[2] ]\n",
    "    # Create a dicitinary where keys are Helices numbers\n",
    "    H_groups = {}\n",
    "    for x in H_status:\n",
    "        if not x[2]:\n",
    "            continue\n",
    "        if x[1] not in H_groups:\n",
    "            H_groups[x[1]]=[]\n",
    "        H_groups[x[1]].append(x)\n",
    "        \n",
    "    # The test is passed if at least two distinct helices feature at least one correctly spaced histidine pair\n",
    "    # ie : if the helice dictionary has more than 1 entrie\n",
    "    HisTestBool = True if len(H_groups) > 1 else False     \n",
    "    return HisTestBool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredBiHistidine=filteredNbHelix.filter(filter_bi_histidine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteins with bi-histidine in helix : 5972\n"
     ]
    }
   ],
   "source": [
    "print(\"Proteins with bi-histidine in helix :\",len(filteredBiHistidine))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter eukaryotic sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Get taxid for entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def function_get_taxid(entry):\n",
    "    reTaxid=re.compile(\"[.]*OX=([\\d]*)\")\n",
    "    taxid=reTaxid.findall(entry.fasta.header)[0]\n",
    "    return taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for e in filteredBiHistidine: \n",
    "    e.get_taxo(function_get_taxid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi=NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_eukaryotic_header(entry):\n",
    "    try:\n",
    "        lineage=ncbi.get_lineage(entry.taxo.taxid)\n",
    "        lineage_rank=ncbi.get_rank(lineage)\n",
    "        superkingdom=[taxid for taxid in lineage_rank if lineage_rank[taxid]=='superkingdom']\n",
    "        if superkingdom:\n",
    "            name=ncbi.get_taxid_translator(superkingdom)[superkingdom[0]]\n",
    "            if name == \"Eukaryota\":\n",
    "                return False\n",
    "            else:\n",
    "                return True\n",
    "        else: \n",
    "            return False\n",
    "    except : \n",
    "        return False\n",
    "        print(taxid,\"not found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredEukaryotic=filteredBiHistidine.filter(filter_eukaryotic_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non-eukaryotic : 851\n"
     ]
    }
   ],
   "source": [
    "print(\"Non-eukaryotic :\", len(filteredEukaryotic))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<pyproteinsExt.topology.TopologyContainer object at 0x118312a20>\n",
      "data structure saved to /Volumes/arwen/mobi/group/NOX_CH/pickle_saved/NOX_annotation_filteredAll_20190613-163034.pickle\n"
     ]
    }
   ],
   "source": [
    "save(filteredEukaryotic,\"filteredAll\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter evalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_evalue(entry,**kwargs):\n",
    "    threshold=kwargs[\"threshold\"]\n",
    "    conserve=0\n",
    "    domains=set()\n",
    "    for h in entry.hmmr: \n",
    "        if float(h.hit.iEvalue) <= threshold:\n",
    "            domains.add(h.domain)\n",
    "    if len(domains)==3:\n",
    "        return True\n",
    "    return False     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_evalue_hit(hit,**kwargs):\n",
    "    threshold=kwargs[\"threshold\"]\n",
    "    if float(hit.hit.iEvalue)<=threshold: \n",
    "        return True \n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredEvalue_3=filteredEukaryotic.filter(filter_evalue,threshold=1e-3).filter_hit(filter_evalue_hit,threshold=1e-3)\n",
    "filteredEvalue_2=filteredEukaryotic.filter(filter_evalue,threshold=1e-2).filter_hit(filter_evalue_hit,threshold=1e-2)\n",
    "filteredEvalue_1=filteredEukaryotic.filter(filter_evalue,threshold=1e-1).filter_hit(filter_evalue_hit,threshold=1e-1)\n",
    "filteredEvalue_4=filteredEukaryotic.filter(filter_evalue,threshold=1e-4).filter_hit(filter_evalue_hit,threshold=1e-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<pyproteinsExt.topology.TopologyContainer object at 0x118321ef0>\n",
      "data structure saved to /Volumes/arwen/mobi/group/NOX_CH/pickle_saved/NOX_annotation_filteredEvalue_3_20190613-163051.pickle\n"
     ]
    }
   ],
   "source": [
    "save(filteredEvalue_3,\"filteredEvalue_3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "dic_evalue={'all':filteredEukaryotic,'1e-1':filteredEvalue_1,'1e-2':filteredEvalue_2,'1e-3':filteredEvalue_3,'1e-4':filteredEvalue_4}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins all : 851\n",
      "Number of proteins 1e-1 : 412\n",
      "Number of proteins 1e-2 : 311\n",
      "Number of proteins 1e-3 : 169\n",
      "Number of proteins 1e-4 : 80\n"
     ]
    }
   ],
   "source": [
    "for evalue in dic_evalue: \n",
    "    print(\"Number of proteins\",evalue,\":\",len(dic_evalue[evalue]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Search with regular expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_seq_motif(entry,**kwargs):\n",
    "    motif=kwargs[\"motif\"]\n",
    "    if motif.search(entry.fasta.seq):\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "NADPH=re.compile(\"G[ISVL]G[VIAF][TAS][PYTA]\")\n",
    "FAD=re.compile(\"H[PSA]F[TS][LIMV]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "== all\n",
      "NADPH : 677 / 851 - 79.55 %\n",
      "FAD : 602 / 851 - 70.74 %\n",
      "Both : 451 / 851 - 53.0 %\n",
      "\n",
      "== 1e-1\n",
      "NADPH : 383 / 412 - 92.96 %\n",
      "FAD : 345 / 412 - 83.74 %\n",
      "Both : 322 / 412 - 78.16 %\n",
      "\n",
      "== 1e-2\n",
      "NADPH : 293 / 311 - 94.21 %\n",
      "FAD : 261 / 311 - 83.92 %\n",
      "Both : 245 / 311 - 78.78 %\n",
      "\n",
      "== 1e-3\n",
      "NADPH : 165 / 169 - 97.63 %\n",
      "FAD : 145 / 169 - 85.8 %\n",
      "Both : 142 / 169 - 84.02 %\n",
      "\n",
      "== 1e-4\n",
      "NADPH : 80 / 80 - 100.0 %\n",
      "FAD : 66 / 80 - 82.5 %\n",
      "Both : 66 / 80 - 82.5 %\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for evalue in dic_evalue: \n",
    "    container=dic_evalue[evalue]\n",
    "    motifNADPH=container.filter(filter_seq_motif,motif=NADPH)\n",
    "    motifFAD=container.filter(filter_seq_motif,motif=FAD)\n",
    "    motifBoth=container.filter(filter_seq_motif,motif=NADPH).filter(filter_seq_motif,motif=FAD)\n",
    "    print(\"==\",evalue)\n",
    "    print(\"NADPH :\",len(motifNADPH),\"/\",len(container),\"-\",round(len(motifNADPH)/len(container)*100,2),'%')\n",
    "    print(\"FAD :\",len(motifFAD),\"/\",len(container),\"-\",round(len(motifFAD)/len(container)*100,2),'%')\n",
    "    print(\"Both :\",len(motifBoth),\"/\",len(container),\"-\",round(len(motifBoth)/len(container)*100,2),'%')\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save NOX predicted proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfasta=filteredEvalue_3.proteins_mfasta()\n",
    "o=open(\"/Volumes/arwen/mobi/group/NOX_CH/predicted_NOX_proteins_global_profile_1e-3.mfasta\",'w')\n",
    "o.write(mfasta)\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Refine Pfam domains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Re-annotate by using new Pfam domains, created with only prokaryote domain sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Write domain sequences in multifasta, from filtered proteins at 1e-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfasta_nad=filteredEvalue_3.get_domain_mfasta(\"PF08030_full\")\n",
    "mfasta_fad=filteredEvalue_3.get_domain_mfasta(\"PF08022_full\")\n",
    "mfasta_ferric=filteredEvalue_3.get_domain_mfasta(\"PF01794_full\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "nad_output=\"/Volumes/arwen/mobi/group/NOX_CH/RefineProfile/nad_binding_proka_1e-3.mfasta\"\n",
    "fad_output=\"/Volumes/arwen/mobi/group/NOX_CH/RefineProfile/fad_binding_proka_1e-3.mfasta\"\n",
    "ferric_output=\"/Volumes/arwen/mobi/group/NOX_CH/RefineProfile/ferric_reduct_proka_1e-3.mfasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "o=open(nad_output,\"w\")\n",
    "o.write(mfasta_nad)\n",
    "o.close()\n",
    "o=open(fad_output,\"w\")\n",
    "o.write(mfasta_fad)\n",
    "o.close()\n",
    "o=open(ferric_output,\"w\")\n",
    "o.write(mfasta_ferric)\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create new hmm profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```console\n",
    "WORKDIR=/mobi/group/NOX_CH/RefineProfile\n",
    "module load clustalw\n",
    "module load hmmr\n",
    "\n",
    "cd $WORKDIR \n",
    "for p in fad_binding nad_binding ferric_reduct; do \n",
    "    clustalw2 -INFILE=$p\\_proka_1e-3.mfasta -OUTFILE=$p\\_1e-3_multiple_alignment.clustal;\n",
    "    hmmbuild -n $p\\_prokaryotes $p\\_prokaryotes.hmm $p\\_1e-3_multiple_alignment.clustal;\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Search NOX proteins with refined pfam domains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Prepare environment\n",
    "```console \n",
    "WORKDIR=/mobi/group/NOX_CH\n",
    "mkdir $WORKDIR/data/refined_profiles\n",
    "cp $WORKDIR/RefineProfile/fad_binding_prokaryotes.hmm $WORKDIR/data/refined_profiles/fad_binding.hmm\n",
    "cp $WORKDIR/RefineProfile/nad_binding_prokaryotes.hmm $WORKDIR/data/refined_profiles/nad_binding.hmm\n",
    "cp $WORKDIR/RefineProfile/ferric_reduct_prokaryotes.hmm $WORKDIR/data/refined_profiles/ferric_reduct.hmm\n",
    "mkdir $WORKDIR/NOX_refined_profiles\n",
    "```\n",
    "\n",
    "* Run hmm and tmhmm\n",
    "```console\n",
    "$SCRIPT_DIR/runHMMR_slurm.sh $WORKDIR/volumes $WORKDIR/NOX_refined_profiles $WORKDIR/data/refined_profiles\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataDir_refined=glob.glob('/Volumes/arwen/mobi/group/NOX_CH/NOX_refined_profiles/uniprot_trembl_v*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 / 293\n",
      "10 / 293\n",
      "20 / 293\n",
      "30 / 293\n",
      "40 / 293\n",
      "50 / 293\n",
      "60 / 293\n",
      "70 / 293\n",
      "80 / 293\n",
      "90 / 293\n",
      "100 / 293\n",
      "110 / 293\n",
      "120 / 293\n",
      "130 / 293\n",
      "140 / 293\n",
      "150 / 293\n",
      "160 / 293\n",
      "170 / 293\n",
      "180 / 293\n",
      "190 / 293\n",
      "200 / 293\n",
      "210 / 293\n",
      "220 / 293\n",
      "Warning:: >> tr|A0A1E7XG60|A0A1E7XG60_9LACO  Serine-rich adhesin for platelets OS=Lactobacillus sunkii OX=481719 GN=sraP PE=4 SV=1\n",
      "   [No individual domains that satisfy reporting thresholds (although complete target did)]\n",
      "\n",
      "\n",
      "230 / 293\n",
      "240 / 293\n",
      "250 / 293\n",
      "260 / 293\n",
      "270 / 293\n",
      "280 / 293\n",
      "290 / 293\n"
     ]
    }
   ],
   "source": [
    "dataContainer_refined=parsing_results(dataDir_refined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins : 190820\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of proteins :\",len(dataContainer_refined))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<pyproteinsExt.topology.TopologyContainer object at 0x11840dda0>\n",
      "data structure saved to /Volumes/arwen/mobi/group/NOX_CH/pickle_saved/NOX_annotation_refined_All_20190613-164854.pickle\n"
     ]
    }
   ],
   "source": [
    "save(dataContainer_refined,\"refined_All\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins with 3 domains :  14483\n",
      "Number of proteins after helix filter :  5057\n",
      "Number of non-eukaryotic proteins : 990\n",
      "Number of non-eukaryotic proteins after evalue 1e-3 filter : 946\n"
     ]
    }
   ],
   "source": [
    "#3 domains \n",
    "filterThreeDomains_refined=dataContainer_refined.filter(filter_three_domains)\n",
    "print(\"Number of proteins with 3 domains : \",len(filterThreeDomains_refined))\n",
    "\n",
    "# Helix filter\n",
    "filterHelix_refined=filterThreeDomains_refined.filter(filter_nb_helix,min_helix=2,max_helix=7).filter(filter_bi_histidine)\n",
    "print(\"Number of proteins after helix filter : \",len(filterHelix_refined))\n",
    "\n",
    "# Eukaryotic filter \n",
    "for e in filterHelix_refined: \n",
    "    e.get_taxo(function_get_taxid)\n",
    "filterNoEukaryotic_refined=filterHelix_refined.filter(filter_eukaryotic_header)\n",
    "print(\"Number of non-eukaryotic proteins :\",len(filterNoEukaryotic_refined))\n",
    "\n",
    "#Evalue 1e-3 filter \n",
    "filterEvalue3_refined=filterNoEukaryotic_refined.filter(filter_evalue,threshold=1e-3).filter_hit(filter_evalue_hit,threshold=1e-3)\n",
    "print(\"Number of non-eukaryotic proteins after evalue 1e-3 filter :\",len(filterEvalue3_refined))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<pyproteinsExt.topology.TopologyContainer object at 0x1ccace438>\n",
      "data structure saved to /Volumes/arwen/mobi/group/NOX_CH/pickle_saved/NOX_annotation_refined_filterAll_20190613-165505.pickle\n",
      "<pyproteinsExt.topology.TopologyContainer object at 0x18b8dd550>\n",
      "data structure saved to /Volumes/arwen/mobi/group/NOX_CH/pickle_saved/NOX_annotation_refined_filter1e-3_20190613-165507.pickle\n"
     ]
    }
   ],
   "source": [
    "save(filterNoEukaryotic_refined,\"refined_filterAll\")\n",
    "save(filterEvalue3_refined,\"refined_filter1e-3\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save predicted proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfasta=filterEvalue3_refined.proteins_mfasta()\n",
    "o=open(\"/Volumes/arwen/mobi/group/NOX_CH/predicted_NOX_proteins_refined_profile_1e-3.mfasta\",'w')\n",
    "o.write(mfasta)\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check regular expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NADPH : 745 / 946 - 78.75 %\n",
      "FAD : 678 / 946 - 71.67 %\n",
      "Both : 507 / 946 - 53.59 %\n"
     ]
    }
   ],
   "source": [
    "motifNADPH=filterEvalue3_refined.filter(filter_seq_motif,motif=NADPH)\n",
    "motifFAD=filterEvalue3_refined.filter(filter_seq_motif,motif=FAD)\n",
    "motifBoth=filterEvalue3_refined.filter(filter_seq_motif,motif=NADPH).filter(filter_seq_motif,motif=FAD)\n",
    "print(\"NADPH :\",len(motifNADPH),\"/\",len(filterEvalue3_refined),\"-\",round(len(motifNADPH)/len(filterEvalue3_refined)*100,2),'%')\n",
    "print(\"FAD :\",len(motifFAD),\"/\",len(filterEvalue3_refined),\"-\",round(len(motifFAD)/len(filterEvalue3_refined)*100,2),'%')\n",
    "print(\"Both :\",len(motifBoth),\"/\",len(filterEvalue3_refined),\"-\",round(len(motifBoth)/len(filterEvalue3_refined)*100,2),'%')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare before and after refined profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "restore a annotated container of  169 elements\n",
      "restore a annotated container of  946 elements\n"
     ]
    }
   ],
   "source": [
    "filterEvalue3=load(\"NOX_annotation_filteredEvalue_3_20190613-163051.pickle\")\n",
    "filterEvalue3_refined=load(\"NOX_annotation_refined_filter1e-3_20190613-165507.pickle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_before=set([p.split(\"|\")[1] for p in filterEvalue3.entries])\n",
    "proteins_after=set([p.split(\"|\")[1] for p in filterEvalue3_refined.entries])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_venn import venn2\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWkAAAD7CAYAAACoomWyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XeYHNWV9/HvmZ6skVDOCWWRDYi0IhgwmGDCOpCMAZvkXV7C2gYWsJvGgI3z7tpgljXIRMOaxbCAyWkBgwgCSyCRRwwoS6MwOZ33j3tHarUmdI+mu6q6z+d55pmZ6uqqU91Vv759K4mqYowxJpyKgi7AGGNM9yykjTEmxCykjTEmxCykjTEmxCykjTEmxCykjTEmxEIX0iIyT0SuDWjeIiK3iUitiMwPooaeiMhpIvJEFqZ7oojUiEidiHyhv6ffw3wni4iKSPF2TqdORKb08Hi1iBy+PfNImtZzInJ2GuNt17L5507ry3MznM+ZIvJiP05vlIi8ICKbROSXInKFiPxXf00/ZV799r6mMa9ul6u/1uPu9DpREakGKoApqlrvh50NfFNVD8lGUQGaC3wJGN+5rNkkIlcD01T1m+mMr6p3AXdloZRfABeo6oNZmHbWqWpV598iMg/4TFWvCq6ignYusAYYpPl1EkZgy5VuS7oYuCibhWSDiMQyfMokoLo/Atq3ykP3TaUbk4B3gi7ChFuaLcVJwLtRCujQL5eq9vgDVAOXA+uAwX7Y2cBz/u/JgALFSc95Djjb/30m8BLwa2A98DFwgB9eA6wCzkh67jzg98CTwCbgeWBS0uOz/GPrgPeAb6Q89ybgUaAeOLyL5RkLPOSf/yFwjh/+HaAJaAfqgEQXz+1clv8ANgBLgMNSlvs6P04jMK2H+X0ZaAFa/fze9sN3AP4ALAc+B64FYknzfzFpfgqcD3wA1AK/A8Q/Ns2/dhtwLYB7u1ieMj9v9a/XR374bL8s63HhfVxX720faorhWu1r/Hrwz6SsO0nTOQv436T/PwTuS/q/BtgjaZ7TcK2dVv+61nU+H7cOfx/4u3897gXKu1nfY8AvfY2fABck18jW63YRcBWwFLce3w7skLJdnAss8+/n95Lmsw/wN/8aLwd+C5SmvI7TuqnxLGAxbvv4GDgv6bFDgM+A7/malgNnJT0+DLc+bgTmAz9Ofv9S5tO5DN8BPgVe8MP3A172tb8NHJK0/SW//ocDVwN3pkzvDD+9NcCVSfMrwmXNR8Ba4D5gaNLjp/vXei1wpX9ft9nG08wRxa1/HwCf+GEHAK/5deQ14IAMl6tzHelpG+51u9xmWdIM6cOB/wGu7WNIt+FWrJgv+FPcxlsGHOFfxKqkF2QTcJB//N/wKxEwALdxnoVr3e/pF3TnpOduAP7Bv+HbbIj+BboRKAf2AFbjg5aUwOkmpNuAS4AS4CQ/v6FJy/0psLOvr6SX+W1+o5Pm8RfgZr+sI3Eb0nk9BOLDwGBgop/2l/1j9+BW5CI/77k9LNfmQPA1fwhcAZQCh/r3Y2YGId1dTefjPtgmAEOBZ+k+pKfgQqAIGIPbOD9PeqwWKOqi/nn49TRlHZ6P+8Acigu487t5Lc4H3gXGA0OAp+g+pL/tX6spQBVuG7kjZbu4x7+Xu/rX4nD/+F64sCv24y4GLk4zpI8BpgICHAw0AHsmhXQbcI1/L4/2jw/xj/8JF34DgF1wIdJbSN/ux68AxuFC8mj/3nzJ/z+iq9efrsPsFj+t3YFmYLZ//GLgFf/al+G2g3v8YzvhArIzF37ll7OnkO4yR5Je3yf9+lDhf9fiPgiKgVP8/8MyWK7OdaSnbTjt7bIvIb0LLpBGkHlIf5D02K5+/FFJw9aypVU0D/hT0mNVuNbtBFwo/l9KfTcD8aTn3t7Dskzw0xqYNOwnwLwMQnoZvmXoh80HTk9a7msymN/mN9r/Pwq30lYkDTsFeLaHQJyb9P99wOX+79uB/8T1r/f2HieH3IHACnwAJq1YV6e+t32o6RmSwhH3Ad1lSPvHa3AfxCf7ZZmP+yZ1FvBQN/XPo+uQ/mbS/z8Dft/NPJ9h65bp4XQf0k8D/5Q07kxci6szeBWYlTLfP3Qz34uBB7papjTev78AF/m/D8F9i0veHlfhPhBivr7kmq6n95CekjTsMvwHUdKwx/HfhlNff7oOs/FJj88HTvZ/L2brb6Zjkl7PH7F1LgzAtWx7CukucyTp9T006fHTgfkp0/gbcGYGy1VM79tw2ttl50/afaaqugjXQro83eckWZn0d6OfXuqwqqT/a5LmW4frKhiL6xfaV0TWd/4ApwGju3puF8YC61R1U9KwpbjWQbo+V/9qJz1/bDfzz3R+k3Ctn+VJy3cz7tO4OyuS/m5gy+t4Ka6lNV9E3hGRb/cwjWRjgRpV7Uiz5kxqGsvWr8/SXqbzPC50DvJ/P4drOR7s/89EdzWlSq2xt/UpeRmWsmVD7er5m9cVEZkhIg+LyAoR2YgLy+E9LoEnIkeJyCsiss6vI0enPHetqrYl/d+5vCN8fZm8B6nLMAn4eso2OBcXqOnq7r2YBDyQNN3FuGAdRcr7om6/0dp0607Jka6WK/W9hMzX+85l6Gkbzni7zHTHVhw4h60L79zJVpk0LDk0+2JC5x8iUoX7KrIM96I+r6qDk36qVPW7Sc9VurcMGCoiA5OGTcR95UvXOBGRlOcv62b+vc0vtdYa3Kfw8KTlG6SqO2dQn5uw6gpVPUdVxwLnATemeUjXMmBCyk7P5Jrr6ft7vZyk99ZPtyedIX2g//t5eg/pnt7/dCzHfd3uNKG7EXGv1aSk/yfivoInN0BSl7dzXbkJ1/UzXVUH4bqXkterLolIGXA/rm9/lKoOxu2D6fW5uO6Wti5q6k3ya1qDa0knb4MDVPWnaUynNzXAUSnTLlfVz0lZd0SkEte/3pPucqRT6raa/F5C5tnQuQzdbsN92S4zCmlV/RC30+XCpGGr/YJ8U0Ri/pNhakaLta2jRWSuiJTidmy8qqo1uJb8DBE5XURK/M8cEZmdZv01uB0ePxGRchHZDbdTJJPD2kYCF/p5fx23k+3RPs5vJTC5MxBVdTnwBPBLERkkIkUiMlVEDs6gPgBE5Osi0hk2tbgVsj2Np76KC+JL/TIeAnwF15cJ8BbwjyJS6Veu72RQ1n241268iAyh929lzwNfxH11/Az4P9wO12HAgm6esxLXR9xX9wEXicg4ERmM+3rfnXuAS0RkRx8C1+N2BCW3Yn/oX6udcd009/rhA3E77+pEZBaQ3NDoSSmuj3U10CYiR+G6jXqlqu24fvOrfU074XbiZeJO4CsicqTf3stF5JCkdW17/B64TkQmAYjICBE53j/2Z+DYpFy4ht7zq7sc6cqjuGw5VUSKReQkXD/4w5ksQG/bcF+2y74cInYNrj8o2TnAD3BfP3bGBdP2uBvXal+H28FyGoDvNjgC10e5DPe16QbcSpuuU3B9SMuAB3D92U9m8PxXgem4HZbXAV9T1Z6+dvU0v//2v9eKyJv+72/hNsR3cW/in8nsq2SnOcCrIlKH25t/kap+0tuTVLUFOA44CreMNwLfUtUlfpRf4/oCVwJ/JLMPuFtw/ZdvA2/iAqOnWt7H7Sz6P///RtzRDC/5wOnKH4Cd/FfNv2RQW3KNT+COBFmA23jb6HpDuhW4A3gBdyRIE/D/UsZ5Hrdz8WngF6raeTLS94FTcTu3bmFLePfIbwMX4j5Mav00Hkpv0QB3tEoVbtuZB9yWwXM7Gx7H41r+q3Etxx/QPyfG/RtuWZ4QkU24nYj7+vm+gzsa425cq7oWdxRLT7rMka74bfhY3FExa3HdEseq6po+LEdP23DG22XnoVEmDSJyJm6n0dygazG54Vuqv1fV1K/CJsTy6aSmqJxsYUxOiEiFiBztv/KOw7XEHgi6LlO4LKSN2ZoACdzX1AW4Iwx+FGhFpqBZd4cxxoSYtaSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbELKSNMSbEioMuwJhMSEKKgQr/UwmUAyW4exPifwvQAbT7nw6gCWgA6jWujTku25g+s3scmlCRhBQBg4Fh/mcQW0K5AhfI26sDqPc/DUl/rwNWa1yb+2EexvQLC2kTGN8q7gzj4f73UCAWZF3ARmB10s8ajWtrsCWZQmUhbXJKEjICmOh/hrOlmyLMFFiPC+xPgRoLbZMrFtImqyQhpcA4XChPwPUjR10H8DmwFKjWuDYEXI/JYxbSpt9JQiqAacAkYDT5fxTRKrYEdm3QxZj8YiFt+oXf4TcRmIlrMed7MHenFlgMfGA7IE1/sJA220USMgDYCRfO+dCV0V/agI+BxRrXlUEXY6LLQtr0iSRkDLALrkujUFvN6VoFLAQ+0bh2BF2MiRYLaZMRSchYYA4wKuhaIqgeWAS8o3FtC7oYEw0W0iYtkpCRuHAeF3QteaABeB14T+O2AZqeWUibHklChuLCeVLQteSh9cB8jWt10IWY8LKQNl2ShOwA7A1MDbqWArACF9Yrgi7EhI+FtNmKP1V7b9xOQdshmFvVuLBeH3QhJjwspM1m/oiNg4Adgq6lgHUAbwBvWX+1AQtpA0hCSoB9ccc7m3BYBTxnrWpjIV3gJCETgAOBqqBrMdtoB14DFlqrunBZSBcoSUgZsD8wI+haTK9W4FrVG4MuxOSehXQB8sc8fwkYEHQtJm1tuJ2Ki4IuxOSWhXSBkYTMBOYS/IX1Td98AjxrZywWDgvpAuGvUncAtnMwH6wBntC41gVdiMk+C+kC4K/v/CXctZ1NfmjABfWqoAsx2WUhnef87aqOwPqf81E78ILG9YOgCzHZYyGdxyQh03Enp1j/c357G7dT0TbmPGQhnackIbviDrEzhWEp8LTtUMw/FtJ5SBKyF7BX0HWYnFsB/NXuZJ5fLKTzjCRkP2C3oOswgVmJC+qWoAsx/cNCOo9IQg7AXb3OFLbVwKN2I9z8YJeizBOSkP2xgDbOCOAof+EsE3EW0nnAd3HsGnQdJlRG4oK6OOhCzPaxkI44ScgeWB+06dpo4EhJiB2CGWEW0hEmCZkM7BNwGSbcxgFfDLoI03cW0hElCRkOHBp0HSYSpkhCvhB0EaZvLKQjSBJSCRwJWH+jSdfekpCJQRdhMmchHTF+R9CR2LU4TGYEOFQSMjjoQkxmLKSj5xDcIVbGZKoUtyOxNOhCTPospCNEErI3MCXoOkyk7QAcLgmRoAsx6bGQjghJyHhgz6DrMHlhPO7u8CYCLKQjwH89PTjoOkxe2U0SMinoIkzvLKSjYS62o9D0vwP9XeNNiFlIh5wkZAowLeg6TF6qxK45HnoW0iHmj4eeG3QdJq/NsOOnw81COtwOAsqDLsLkvYPssLzwspAOKUnILMBaOCYXKoEDgi7CdM1COoR8N4f1FZpcmiEJmRB0EWZbFtLhtDdgF2w3uXaQ3SggfCykQ0YSMgSYGXQdpiANwG4eEToW0uGzL+5iOMYEYXdJiO2sDhEL6RCRhIzFdhaaYJUAdu3pELGQDpf9gi7AGGAnSUhV0EUYx0I6JCQh04HhQddhDBDD7bw2IWAhHQL+RqFzgq7DmCTT/U5sEzAL6XDYCbCvlyZMBGs4hIKoatA1FDR/8fVTyEFIi6JD2mmPKVqk7hO6A2gsoqg+RlG72FElZhsPaFxXB11EIbMbmQZvClkI6EFttE9tpnVaEx1Tm2BiC7HRLZSU9PCet0HHmhJaPyinfVEFsrCSWE0Zdk2HwrYb8HTQRRQya0kHTBJyIv1wz8KSDjoOqKP5sA3o9CZKqzr65wO4WWhfWkbLknI65lcRe3uAXfCpwHQAd2tcG4IupFBZSAdIEjIaOG57pjG7geYTa2nfs56yMiXWT6V1a22MlicH0/rIYMrWF9s3sQKxQOP6WtBFFCoL6QBJQg6lDxf0L+mg4/ANNB1fS2xcK4HcWaMddH4VjXcMp9i6RPJeE3CXxrU96EIKkbWEAiIJqaAPd/7eq46mi1YQG9JOZRbKSlsMZP86KvetQxdV0Pifo4gttbDOV+XAjsCHQRdSiOwQvODMJIPXf1Ab7Vd9RuPVn1M+pD08V8grAtmtkYrfVFN86hoaRLGvZvlpVtAFFCoL6eDMTnfEo2ppuOVj2LeeimwWtD2KoeiUtVT+ezXNY1poDboe0+/GSkIGBV1EIbKQDoAkZAQwsLfxxrTQ+qtqmv5pFZWVOdgp2B8mt1D+22pix6/DjgbIP9aaDoCFdDB27G2EOXU0/raa2PTm6B3yVqoUnb2ayp8vpXFoG21B12P6Tcb7UMz2s5AOxuSeHjymloYrP6e8VKP9/sxqouLfq9FxzbQEXYvpF4Pseh65F+kQiCK/kg/u7vHTV1N//ioqY3ly4f8d2in55acU7dhkQZ0nJgVdQKGxkM69yd09cMZq6r+xjgE5rCUnBnRQ/LNPiU1tojnoWsx2s5tS5JiFdO512R99+mrqv5aHAd2pXIldV0PMjvyIvFF2e63cspDOIX+3i20u7H9ULQ352IJO5VvU2M7ESBOsNZ1TFtK5tU0rekwLrWevjt4RHH01uJ2SGz51l0sNuhbTZ9YvnUMW0rk1PvmfmKI//JyOqB/FkanRrZR9e5UdRx1h4yUhBbXOBsle6NwamfzP2atomNASzAWSgnbMeiptR2JklQBjgi6iUFhI54gkZAfYEsi719N09PpgL5IUpBjIpcvct4mgazF9MrL3UUx/sJDOnc0rdVU77T9YTqwoT46F7quxrZR9c411e0SU3dk+Ryykc2dzSF+4guYdQnQluyCdsI6KSXZGYhRt992ETHospHNnJLijOfatC+/V7HKtGIouW2ZdHhFUJQkpyP0puWYhnQOSkBgwDOCs1bQVejdHqgktlM2pozHoOkzGrDWdAxbSuTEMKBrVQus+dYVzTHQmTl4bdAWmD6xfOgcspHNjJMA31tGSLxdO6m8zmqiwizBFjoV0DlhI58bgsg46DtporeiefHWdnS4eMRbSOWAhnRtVh22guTwid1cJyr51lJV00BF0HSZtgyQhdpRSlllI50bVERusm6M35Ups7iaagq7DZKRgT8jKFQvpHIgplRObKQ26jig40j7MosZCOssspLNMElIyoxEpsdc6LTs22YdZxFhIZ5kFR/YN+EID7UEXERWVSmxEq90YIEIspLPMQjr7qnZrsK/wmZjdaEd5RIiFdJZZSGdflX2Fz8ysRjtNPEIspLPMQjrLDtpIeaUdepeRaU22XkaIhXSW2caQZYdtKMyL+m+PiS0UB12DSZuFdJbZxpBFIvLlkhg3lxRRsetIXrzqIB7L5fwvfowzajaya1mMTXd/lUTn8J+/zBffWMYXi4SOyYNZeP1h3F/fQuzKZ/jm6nomIehXZ3PvP87m/VzW22lAB8XDWmlfWxKxbyBLGMVDnLv5/yaGM5uHWMUU6hkNQBsVFNPIpfyYR9iHdzhy8/gNjONErmV3Pst57X1nV3TMMgvpLBGRGPC7M+Zw2eHjGHLuw1zx4qe8PXciy3NVwyGTebmqlGdvXcBZncPuX8zMJWvY47+O45qqUto+qmUgwE2vcyDAXV/lmo9qGXj1c1x43EyuLy4Kpn94bCutkQvpWaxkFj8GoA3hBn7G3ixgR57ePM5tfI1Sf8W/Y5jPMcwHYCHjeJh/ilhAg30bzzp7gbNnH+DDI6ewekAp7bOH89pz1eyeywJOmMUHwyupTx72zMccfORU/lpV6o6gmDqETQAr6hgzcxhLOoeVxmh46uPg7gpdGvWTw19iNuWsZkfWbR7WASxnb/bmtW3GX8AcxnUxPPwsQ7LMXuDsGQfUlKv7tjK0gtoNzQwJuCY2NDPqndVM/9YDXH7Wg3z/8Q9dEI8fxGcLV7F7cxtFC5YzbF0jk5ZtCq7e0qjf+3AJc5jsW8mdXmU6JWxkJqu2Gf9z9mbPlPGjwTIky+wFzh53bLRudYx04MHToRQ1tlI57wR+etLO/PkPCzivQ+G7e/PSwDLWf+chrrz5DU4aXslHMQnuYkflHcG/Vn3WRIzV7Mb+vLHV8EXMYVIXreXX2JEYLezCslyV2I/sHIAssz7p7PkMmND5z7pGhuxQxvoA6wFgQCm1+41nQZHAl6dRfdtbdCxdT9WOQ6i74XDu6xzvjL9w2dShXbT4cqRDIrzxv8guVPEpY11XEgCtFLGSPTmCa7cZfyFzGB/Jrg6TA9aSzp7XgOnPL2VEfQuxxWuYc/Bk3g66qNnDeeutFcwCePUzRnYoxZMGU1fbSOnaBnfSzb3vMLtIaM/lTs5UzVEO6ffZhykpofsSs6lkBZNSPqjbEZazF/tENqSjvvcg9KwlnSWq2iYiF/zHfG6OFVG28wheOjDHoXfBo5y9oo4ZLe1Uff2/ueHAiTx03l68dPnTnHHq/cSLhLaTdua2IoGlGxj40xe5SAStLGH9Jftxay5rTdUc1eZDHaWsZTYnc+dWw5cwhx276HN+hemUUcs01uSqxH5mIZ1lohrdrr8ouP10OWhou2u5mvRdOoHmxZV2IlAEtGhc5wVdRD6LanslMjbF7IpufbG0zL7lRYRdDCvLLKSzrLrMVuJM1cZobYhF7ESWwtUYdAH5zkI6yx4ZbHfAzlR1mX37iJCGoAvIdxbSWba4kob11uWRkQ/KI3yMdOGxkM4yC+nsq3+/3EI6E0sqbL2MEAvpLLONIfvq/m4Xc8zIe+V2k4QIsZDOMgvp7Kt7a4AdqZCuDTFaNxbbTsMIsZDOMgvpLNO4diwto61B7Ga06VhUYV1DEWMhnWUW0rlR/14FzUEXEQWPDLF1MmIspLPMNojcqPvfIRG+FkWO1MZoXVhJedB1mLR1wNbXKzf9z0I6N+peq6Ki1g7F69Gzg+yY8ohZr3G1a3dkmYV0bqwD+Kud2NKtNuh4YKhdqyNiVgddQCGwkM6NVQAPDqG82XYgdum1KprWF9tRMBET1Sv3RYqFdG6sB3c9ir8OpinoYsLoruEW0BFkIZ0DFtI5oHFV/FfDe4ZZazrV/AE0LC2zE1giRoG1QRdRCCykc2cVQEOM2INDrDXdqb6Itl+Psb7oCKrVuNoVHnPAQjp3Nt8v8M7hVFaXWlAD3DSK1jq7LGkUWVdHjlhI587mkFZBrhlPrNC7PRZU0vD8ICqCrsP0iYV0jlhI54jGtQGo6/x/dQklN48s3LMQG4T2X1g3R5StCLqAQmEhnVsrk/95cjCV8wcU5mm1t4ykxS6kFFkNGldrSeeIhXRufZo64JdjKCu0mwK8PoCGpwZbN0eELQ26gEJiIZ1bS3HXO9isIUbs+rF0tKYMz1fvl9N47TgL6IizkM4hC+kc0ri2AMtShy+upOy6cbS05XlQ15TSfMUEytrFLjYVYW10sQ6b7LGQzr3qrga+UUX59eNoztegXlZC86UTKW4usnUu4j6346NzyzaY3KuGrm+0+loVFdeOo6VF8iuoq0tpumQSxXY8dF6wro4cs5DOMX8o3sruHn+jivJLJ9JaV0RetFY+KKPp+5MobbCAzhfb7Pw22WUhHYzqnh78qJyyCyejK4uje2nTdtD7hlL/vUmUWRdH3ljtGxkmh2zjCcYnvY2wuoSS86ZQ8sAQGqLWT72ymJZLJtF6xwgGqO0kzCcfBF1AIbKQDoDGdRNJp4l3p12QW0dSeeFk2j4uC/+1PjpAHx1M/XlTKPmk3K5ql2fasZAOhIV0cN5Nd8SaMkovmkz5zSNpbArp9T5qY7ReMYHmm0YxwA6xy0ufaFwL9jIGQbKQDs5HkNm1Ox4eQsU5U9DXB9DQ0c0RIrm2IUbr7cNpOGcKsXfsJrL5bEnQBRQqUQ3Ftl6QJCH7Abv15bljWmj92jpaDt5IeZnm/siJlcW03D+UticGU2Et57y3XuN6X9BFFCq7ZVGw3gV2hcxDbnkpJf8xmpL/HEnHoRtoOHwjMq2J8qI+TCtdrdAxv4qmB4cQW1xJGVi/c4FYFHQBhcxa0gGThBwBTO6PaQ1uo+2wDbTs1gDjWyge1kZJbDtCu0Fory6jZUkFHQsriS2qpLTJDqcrNC3AnXaWYXAspAMmCRkNHJeNaZd00DGxhdZpTbRPa0InNxMb0UpRDESAmCLtgjYL2lyENhbRUVOKvlMBiyopWV5KSTbqMpHyd43rK0EXUcgspENAEnICMDLoOoxJ0Q7cq3Gt63VMkzX21TUc3g66AGO68K4FdPAspENA4/oJaZzcYkwOtQALgi7CWEiHifX7mTD5u8Y19Ge5FgIL6ZDQuK6glwsvGZMjjcDCoIswjoV0uLxKxC6mZPLSAo1rQd13M8wspENE47oBWBx0HaagbSKD68qY7MubMw5FZB7wmapelca4CkxX1Q/7MJ9q4GxVfSrjItPzBjAdKGUZA/kT51HHRCbwAmXUU8dwzuUOPmYYt3M9V/JdSqz1bfrNGxpXW59CJDIhLSInA5cAuwD1uGsy/xG4SfPoYG+Na5Mk5C1gH57lIEqp4youtO88JgdWYZcjDZ1IbPoi8j3g34CfA6OBUcD5wD8Q0etHiEhPH5ALgY3UM5RBLI/Gu2Qirh14TuP50+DJF6FvSYvIDsA1wLdU9f6khxYAp/XwvHOAy4ChwIvA+aqafCv6o0XkYmAQcBtwmap2iMhU4BZgd9zlQB8H/llV16dR6zygCZgK7Ae86ete6h9X4ALgYtxrv6OIHID7AJoBvA9cpKovS5mcSAv7I8A1HMYXuZGlTGcTI/gut24z83VU8N98nTXsAigTeJlTeYhilPcZwcOcQT3jEdoZwhL+mVt6Wx5TUN7QeO/ruMm9KLTR9gfKgAfTfYKIHAr8BPgGMAZ3h+M/pYx2IrA3sCdwPPDtzqf7544FZgMTgKszqPc04MfAcOAt4K6Ux08A9gV2EpGhwCPAvwPDgF8Bj4jIMG3WkxjEY+zI4/yICzmwl+v53sWZFNHBhVzFt7mWVezE/zIXgKc4ntG8wxVcwr9wGXvxbAbLY/LfKuys19CKQkgPB9Zn0ictAAAKsElEQVSobrkKl4i8LCLrRaRRRA7q4jmnAbeq6puq2gz8K7C/iExOGucGVV2nqp8CvwFOAVDVD1X1SVVtVtXVuOA8OIN6H1HVF/x8r/TznZD0+E/8fBuBY4APVPUOVW1T1XtwF1f/CgAbWUMbjb3OcRkDWccunMq9DKSFMWxiJ56imjkAFNFOHcP4jB2opI39yHiHqclb7cDz1s0RXqHv7gDWAsNFpLgzqFX1AAAR+YyuP2jG4roa8OPXichaYBxbThipSRp/qX8OIjIS17I9EBjop1+bQb2bp+vnu85Puyb1cT98acrzl/o6nVUsxR073f0H6jKGocT4FT/fPEwRynzdX+F+HuF47uQKimlgNk9wHC9nsEwmf72hcc1k/TY5FoWQ/hvuNlPHA/f3Mm6nZcCkzn9EZACuO+HzpHEmAO/4vyf654Dr6lBgN1VdKyInAL/NoN7NrWYRqcL1iSf3hSe3WLaqM6mWxzb/10Qj8DqwT7dzHE0tQhuX8y9dHo43jo2cyx0AvMI0HucSZvEBM1id5jKZ/LQa6+YIvdB3d/gddgngRhH5mohUiUiRiOwBDOjmaXcDZ4nIHiJSBlwPvKqq1Unj/EBEhviuiIuAe/3wgUAdsF5ExgE/yLDko0VkroiU4vqmX1XVmm7GfRSYISKnikixiJwE7AQ8nDLe28DKbuc4ng0M5V1u5+vUUk47wgeM4GWmA/AYe7GUwQBUUQ8oMTu2usDZ0RwREYWWNKr6MxH5HLgUuB13nPTHuKM3tvnarqpPi8gPcS3vIX6ck1NGexB34sgOwDzgD354ws9jA/AhcAfu+Ox03Q3EcTs836SHI1B8S/1Y3NEdN/n5Hauqa7YaL64qCXkW3bxzc1uncRv3cyI3kqCDMspZwy6+Rb6CybzBN2inghI2sjv3MpW1GSyTyT/PWzdHNNhF//tRJmc99mn6CZkMHJGNaZuC8rbG9dWgizDpCX13h9lC41oNzA+6DhNpNdg6FCkW0hGjcX0Ld9KLMZnaADxt/dDREok+6ahQ1TNzNKsXcGdKjs7R/Ez0tQCPa1xbgi7EZMZa0hHkr1L2BLAx6FpMJCiuBW2nfUeQhXRE+VsbPYZrIRnTk/ka7/YwUBNyFtIR5ltGT2F3czHde0vjaiesRJiFdMRpXD8DnsaC2mxrkcbVjuSIOAvpPKBx/QR4Bgtqs8USjatdnyUPWEjnCY3rx7igtsOrzBKN6wtBF2H6h51xmGckITsCh2EfwIXqHY3rS0EXYfqPhXQekoSMx50+bsfBF5aFGte/BV2E6V8W0nlKEjIa+DIRvQekyYgCr/mzUU2esZDOY5KQobgW9aCgazFZ0wo8o3FNvXmEyRMW0nlOElKG66MeH3Qtpt9twp3qvS7oQkz2WEgXAEmI4O7ssnvQtZh+sxx40p95avKYhXQBkYRMAQ7BdihG3RLgRX8NF5PnLKQLjO+nPhJ3mzATLR3AKxrXRUEXYnLHQroA+X7qQ0m6aa4JvVrgWY1vfWs1k/8spAuYJGQm7l6MdpheeCnuRsRvaFzbgy7G5J6FdIGThFQCc4HJAZditrUB13peFXQhJjgW0gYASchU4B+A8qBrMSiwCHeCSlvQxZhgWUibzSQh5cABwLSgaylgG4HnNK4rgi7EhIOFtNmGJGQiLqztTMXcaQIWAO9a37NJZiFtuiQJKQJmAXsClQGXk8/acF0bb9lNYk1XLKRNjyQhMWBnYA+sv7o/KfAe7qiN+qCLMeFlIW3SIgkpAXYFdsMO2dteS3E3h60NuhATfhbSJiP+RJg9gJ2AkoDLiRIFPgH+bofUmUxYSJs+kYSUAjNwYT044HLCrAV3rY1FGte6oIsx0WMhbbabvxPMbGASdtuuTquAxcBHdqyz2R4W0qbf+OOsZwAzgSEBlxOERlyXxmKN69qgizH5wULaZIUkZDiuZT0RGBFwOdm0AajG7QxcqXHboEz/spA2WScJqcBdcW8i7g4xUT86ZBU+mO0IDZNtFtImp/xJMqNxgT0GGArEAi2qZx3AOlwwrwZqNK4NwZZkComFtAmUD+3BwDD/M9z/LgugHMVdt3kNW0J5nZ2mbYJkIW1CSRJShQvsKtxp6RUpP5VkfiRJK1DfxU8DUAestyMxTNhYSJvI8ifWVODu2Sj+p1OH/2n3v5s0rq05L9KY7WQhbYwxIWYnHhhjTIhZSBtjTIhZSBtjTIhZSBtjTIhZSBtjTIhZSBvTBRH5roisFJE6ERnmf0/JwnzOFJEX+3u6Pcyv2+USkXkicm2uajHpKQ66AGOyQUSqgVG446TrgMeAC1R7v6aziJQAvwL2U9W3/eCqLJWaM/m6XPnOWtImn31FVatwd5L5AvCvaT5vFO5+ju9kq7BsEJHeGl2RXK5CZyFt8p6qrgAex4U1ACJSJiK/EJFP/df/34tIhYjMwN0gFmC9iDzjx1cRmeb/nicivxORR0Rkk4i8KiJTk6Y9S0SeFJF1IvKeiHwj6bFhIvKQiGwUkfnA5uelEpHJfr7nisgyEVkuIt9LevxqEfmziNwpIhuBM/1y/caPv8z/XZbOcnUx/2NF5C0RWS8iL4vIbkmPXSYin/vlf09EDkvv3TCZspA2eU9ExgNHAR8mDb4Bd4OCPYBpwDjgR6r6Pu7u6ACDVfXQbiZ7CpDA3dzgQ+A6P68BwJPA3cBIP96NItI5zd8BTbgrAH7b//Tmi8B04AjgchE5POmx44E/4y5SdRdwJbCfX67dgX2AqzJYLvxy7AncCpyHu+DVzcBDPvBnAhcAc1R1IHAk7tKtJgsspE0++4uIbAJqcFe1iwOIiADnAJeo6jpV3QRcD5ycwbT/R1Xnq2obLhw7W+nHAtWqepuqtqnqm8D9wNdEJAZ8FfdhUK+qi4A/pjGvhB9/IXAbLvg7/U1V/6KqHaraCJwGXKOqq1R1Ne6D5PQMlqvTOcDNqvqqqrar6h+BZtwHQDvuKoU7iUiJqlar6kd9mIdJg4W0yWcn+JbeIcAs3FX1wN0pphJ4w3+VX4/bsZjJHWRWJP3dwJYdcJOAfTun66d9Gu4a2iNwO+trkp67NI15pY4/tpvH8I8lTzN1/HRNAr6XshwTgLGq+iFwMXA1sEpE/iQifZmHSYOFtMl7qvo8MA/4hR+0Bnc/wp1VdbD/2cHvZNxeNcDzSdMdrKpVqvpd3PWp23Bh12liGtNMHX9Z0v+pV0hbhgvY7sZPVw1wXcpyVKrqPQCqereqzvXzUlz3kckCC2lTKH4DfElE9lDVDuAW4NciMhJARMaJyJH9MJ+HgRkicrqIlPifOSIyW1Xbgf8BrhaRShHZCTgjjWn+0I+/M3AWcG8P494DXCUiI0RkOPAj4M4+LMctwPkisq84A0TkGBEZKCIzReRQESnD9a834rpATBZYSJuC4Ptnbwd+6Addhtvh94o/MuIp3F3Ot3c+m3A7+E7GtWBX4FqZnXeauQDXNbIC17q/LY3JPu9rfRr4hao+0cO41wKvA38HFgJv+mGZLsfruH7p3+LuVvMhcKZ/uAz4Ke4byQrcDtIrMp2HSY9dT9qYkBKRycAnQInfQWkKkLWkjTEmxCykjTEmxKy7wxhjQsxa0sYYE2IW0sYYE2IW0sYYE2IW0sYYE2IW0sYYE2IW0sYYE2L/H6knWRiZMWqAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "v=venn2([proteins_before,proteins_after])\n",
    "v.get_label_by_id(\"A\").set_text(\"Global profiles\")\n",
    "v.get_label_by_id(\"B\").set_text(\"Refined profiles\")\n",
    "plt.title(\"Number of proteins found with global and refined profiles\")\n",
    "plt.savefig(\"/Users/chilpert/Work/NOX_visu_new/nb_prot_before_after_refined_profiles.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All proteins found with global profiles are also found with refined profiles and we have 777 more proteins with refined profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proteins distribution in taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "o=open(\"/Volumes/arwen/mobi/group/NOX_CH/Taxonomy/proteins_taxonomy_taxid_krona.tsv\",\"w\")\n",
    "o.write(\"#protein\\ttaxid\\tscore\\n\")\n",
    "for e in filterEvalue3_refined : \n",
    "    o.write(e.prot+\"\\t\"+e.taxo.taxid+\"\\t1\\n\")\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['ktImportTaxonomy', '/Volumes/arwen/mobi/group/NOX_CH/Taxonomy/proteins_taxonomy_taxid_krona.tsv', '-o', '/Users/chilpert/Work/NOX_visu_new/taxo_krona_filtered.html'], returncode=0)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Run kronaTools\n",
    "subprocess.run(['ktImportTaxonomy','/Volumes/arwen/mobi/group/NOX_CH/Taxonomy/proteins_taxonomy_taxid_krona.tsv','-o','/Users/chilpert/Work/NOX_visu_new/taxo_krona_filtered.html'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Re-refined proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfasta_nad=filterEvalue3_refined.get_domain_mfasta(\"nad_binding_prokaryotes\")\n",
    "mfasta_fad=filterEvalue3_refined.get_domain_mfasta(\"fad_binding_prokaryotes\")\n",
    "mfasta_ferric=filterEvalue3_refined.get_domain_mfasta(\"ferric_reduct_prokaryotes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nad_output=\"/Volumes/arwen/mobi/group/NOX_CH/ReRefineProfile/nad_binding_proka_1e-3.mfasta\"\n",
    "fad_output=\"/Volumes/arwen/mobi/group/NOX_CH/ReRefineProfile/fad_binding_proka_1e-3.mfasta\"\n",
    "ferric_output=\"/Volumes/arwen/mobi/group/NOX_CH/ReRefineProfile/ferric_reduct_proka_1e-3.mfasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o=open(nad_output,\"w\")\n",
    "o.write(mfasta_nad)\n",
    "o.close()\n",
    "o=open(fad_output,\"w\")\n",
    "o.write(mfasta_fad)\n",
    "o.close()\n",
    "o=open(ferric_output,\"w\")\n",
    "o.write(mfasta_ferric)\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```console\n",
    "WORKDIR=/mobi/group/NOX_CH/ReRefineProfile\n",
    "module load clustalw\n",
    "module load hmmr\n",
    "\n",
    "cd $WORKDIR \n",
    "for p in fad_binding nad_binding ferric_reduct; do \n",
    "    clustalw2 -INFILE=$p\\_proka_1e-3.mfasta -OUTFILE=$p\\_1e-3_multiple_alignment.clustal;\n",
    "    hmmbuild -n $p\\_prokaryotes $p\\_prokaryotes.hmm $p\\_1e-3_multiple_alignment.clustal;\n",
    "done\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Prepare environment\n",
    "```console \n",
    "WORKDIR=/mobi/group/NOX_CH\n",
    "mkdir $WORKDIR/data/re_refined_profiles\n",
    "cp $WORKDIR/ReRefineProfile/fad_binding_prokaryotes.hmm $WORKDIR/data/re_refined_profiles/fad_binding.hmm\n",
    "cp $WORKDIR/ReRefineProfile/nad_binding_prokaryotes.hmm $WORKDIR/data/re_refined_profiles/nad_binding.hmm\n",
    "cp $WORKDIR/ReRefineProfile/ferric_reduct_prokaryotes.hmm $WORKDIR/data/re_refined_profiles/ferric_reduct.hmm\n",
    "mkdir $WORKDIR/NOX_re_refined_profiles\n",
    "```\n",
    "\n",
    "* Run hmm and tmhmm\n",
    "```console\n",
    "$SCRIPT_DIR/runHMMR_slurm.sh $WORKDIR/volumes $WORKDIR/NOX_re_refined_profiles $WORKDIR/data/re_refined_profiles\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataDir_re_refined=glob.glob('/Volumes/arwen/mobi/group/NOX_CH/NOX_re_refined_profiles/uniprot_trembl_v*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dataContainer_re_refined=parsing_results(dataDir_re_refined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3 domains \n",
    "filterThreeDomains_re_refined=dataContainer_re_refined.filter(filter_three_domains)\n",
    "print(\"Number of proteins with 3 domains : \",len(filterThreeDomains_re_refined))\n",
    "\n",
    "# Helix filter\n",
    "filterHelix_re_refined=filterThreeDomains_re_refined.filter(filter_nb_helix,min_helix=2,max_helix=7).filter(filter_bi_histidine)\n",
    "print(\"Number of proteins after helix filter : \",len(filterHelix_re_refined))\n",
    "\n",
    "# Eukaryotic filter \n",
    "filterNoEukaryotic_re_refined=filterHelix_re_refined.filter(filter_eukaryotic_header)\n",
    "print(\"Number of non-eukaryotic proteins :\",len(filterNoEukaryotic_re_refined))\n",
    "\n",
    "#Evalue 1e-3 filter \n",
    "filterEvalue3_re_refined=filterNoEukaryotic_re_refined.filter(filter_evalue,threshold=1e-3).filter_hit(filter_evalue_hit,threshold=1e-3)\n",
    "print(\"Number of non-eukaryotic proteins after evalue 1e-3 filter :\",len(filterEvalue3_re_refined))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test refined profiles with evalue 1e-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "filteredEvalue_1=load(\"NOX_annotation_filtered1e-1_20190524-143321.pickle\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Write domain sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mfasta_nad=filteredEvalue_1.get_domain_mfasta(\"PF08030_full\",1e-1)\n",
    "mfasta_fad=filteredEvalue_1.get_domain_mfasta(\"PF08022_full\",1e-1)\n",
    "mfasta_ferric=filteredEvalue_1.get_domain_mfasta(\"PF01794_full\",1e-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nad_output=\"/Volumes/arwen/mobi/group/NOX_CH/RefineProfile/nad_binding_proka_1e-1.mfasta\"\n",
    "fad_output=\"/Volumes/arwen/mobi/group/NOX_CH/RefineProfile/fad_binding_proka_1e-1.mfasta\"\n",
    "ferric_output=\"/Volumes/arwen/mobi/group/NOX_CH/RefineProfile/ferric_reduct_proka_1e-1.mfasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o=open(nad_output,\"w\")\n",
    "o.write(mfasta_nad)\n",
    "o.close()\n",
    "o=open(fad_output,\"w\")\n",
    "o.write(mfasta_fad)\n",
    "o.close()\n",
    "o=open(ferric_output,\"w\")\n",
    "o.write(mfasta_ferric)\n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create new hmm profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```console\n",
    "WORKDIR=/mobi/group/NOX_CH/RefineProfile\n",
    "module load clustalw\n",
    "module load hmmr\n",
    "\n",
    "cd $WORKDIR \n",
    "for p in fad_binding nad_binding ferric_reduct; do \n",
    "    clustalw2 -INFILE=$p\\_proka_1e-1.mfasta -OUTFILE=$p\\_1e-1_multiple_alignment.clustal;\n",
    "    hmmbuild -n $p\\_prokaryotes $p\\_1e-1_prokaryotes.hmm $p\\_1e-1_multiple_alignment.clustal;\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Relaunch NOX proteins search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```console\n",
    "WORKDIR=/mobi/group/NOX_CH\n",
    "mkdir $WORKDIR/data/refined_profiles_1e-1\n",
    "cp $WORKDIR/RefineProfile/fad_binding_prokaryotes.hmm $WORKDIR/data/refined_profiles_1e-1/fad_binding.hmm\n",
    "cp $WORKDIR/RefineProfile/nad_binding_prokaryotes.hmm $WORKDIR/data/refined_profiles_1e-1/nad_binding.hmm\n",
    "cp $WORKDIR/RefineProfile/ferric_reduct_prokaryotes.hmm $WORKDIR/data/refined_profiles_1e-1/ferric_reduct.hmm\n",
    "mkdir $WORKDIR/NOX_refined_profiles_1e-1\n",
    "\n",
    "$SCRIPT_DIR/runHMMR_slurm.sh $WORKDIR/volumes $WORKDIR/NOX_refined_profiles_1e-1 $WORKDIR/data/refined_profiles_1e-1\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataDir_refined1=glob.glob('/Volumes/arwen/mobi/group/NOX_CH/NOX_refined_profiles_1e-1/uniprot_trembl_v*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dataContainer_refined1=parsing_results(dataDir_refined1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "print(\"Number of proteins :\",len(dataContainer_refined1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save(dataContainer_refined1,\"refined_1e-1_All\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3 domains \n",
    "filterThreeDomains_refined1=dataContainer_refined1.filter(filter_three_domains)\n",
    "print(\"Number of proteins with 3 domains : \",len(filterThreeDomains_refined1))\n",
    "\n",
    "# Helix filter\n",
    "filterHelix_refined1=filterThreeDomains_refined1.filter(filter_nb_helix,min_helix=2,max_helix=7).filter(filter_bi_histidine)\n",
    "print(\"Number of proteins after helix filter : \",len(filterHelix_refined1))\n",
    "\n",
    "# Eukaryotic filter \n",
    "filterNoEukaryotic_refined1=filterHelix_refined1.filter(filter_eukaryotic_header)\n",
    "print(\"Number of non-eukaryotic proteins :\",len(filterNoEukaryotic_refined1))\n",
    "\n",
    "#Evalue 1e-3 filter \n",
    "filterEvalue3_refined1=filterNoEukaryotic_refined1.filter(filter_evalue,threshold=1e-3)\n",
    "print(\"Number of non-eukaryotic proteins after evalue 1e-3 filter :\",len(filterEvalue3_refined1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_proteins=set([p.split(\"|\")[1] for p in filterEvalue3_refined1.entries])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not new_proteins.difference(proteins_after):\n",
    "    print(\"No difference between refined profiles with domains 1e-3 and with domains 1e-1\")\n",
    "else: \n",
    "    print(\"Difference between refined profiles with domains 1e-3 and with domains 1e-1. Need to check.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
